### DISCUSSION TITLE
How to render a file link in the blade file

### DISCUSSION THREAD

**Discussion by ctocph** [area17/twill#1705] [General]
Hello,

I've added files to a module etc following https://twill.io/docs/form-fields/files.html#files
So all in that guide is done.

Now, I want to make it show in the blade file...and there is no documentation with examples on how to do that

Tried something like
`<a href="{{ $service->file('name') }}">filename</a>`
but it's not coming

Any ideas?

**Comment by haringsrob** [comment]
Hello @ctocph,

The argument you pass into the file method should be that of your file field name.

So if you use:
```php
@formField('files', [
    'name' => 'single_file',
    'label' => 'Single file',
    'note' => 'Add one file (per language)'
])
```

Then you can get the file using:

```php
$service->file('single_file');
```

Of if you allow multiple files you can use:

```php
$service->filesList('fieldname');
```

**Comment by ctocph** [comment]
Hi, thanks @haringsrob. I think this will work. But I found out the files object is empty. 
I tried this:
1. php artisan twill:make:module -TS testingfiles (to generate a new "clean" module)
2. I add the proper admin.php and twill-navigation items as suggested:
`Add to routes/admin.php:

    Route::module('testingfiles');

Setup a new CMS menu item in config/twill-navigation.php:

    'testingfiles' => [
        'title' => 'Testingfiles',
        'module' => true,
    ],`
3. I go to twill http://admin.something.test:8000/testingfiles/1/edit
4. I add a file, it's uploaded correctly
5. Save
6. Reload - it's not there anymore, gone gone

File can be seen in the files table in the database

Feels like something else needs to be added that the twill:make:module is not adding.
I have (in the auto generated):
Testingfile.php Model
`<?php

namespace App\Models;

use A17\Twill\Models\Behaviors\HasTranslation;
use A17\Twill\Models\Behaviors\HasSlug;
use A17\Twill\Models\Behaviors\HasMedias;
use A17\Twill\Models\Behaviors\HasFiles;
use A17\Twill\Models\Behaviors\HasPosition;
use A17\Twill\Models\Behaviors\Sortable;
use A17\Twill\Models\Model;

class Testingfile extends Model implements Sortable
{
    use HasTranslation, HasSlug, HasMedias, HasFiles, HasPosition;

    protected $fillable = [
        'published',
        'title',
        'description',
        'position',
    ];
    
    public $translatedAttributes = [
        'title',
        'description',
        'active',
    ];
    
    public $slugAttributes = [
        'title',
    ];
    
    public $mediasParams = [
        'cover' => [
            'default' => [
                [
                    'name' => 'default',
                    'ratio' => 16 / 9,
                ],
            ],
            'mobile' => [
                [
                    'name' => 'mobile',
                    'ratio' => 1,
                ],
            ],
            'flexible' => [
                [
                    'name' => 'free',
                    'ratio' => 0,
                ],
                [
                    'name' => 'landscape',
                    'ratio' => 16 / 9,
                ],
                [
                    'name' => 'portrait',
                    'ratio' => 3 / 5,
                ],
            ],
        ],
    ];
}
`
and in Repositories TestingfileRepository.php:
`<?php

namespace App\Repositories;

use A17\Twill\Repositories\Behaviors\HandleTranslations;
use A17\Twill\Repositories\Behaviors\HandleSlugs;
use A17\Twill\Repositories\Behaviors\HandleMedias;
use A17\Twill\Repositories\Behaviors\HandleFiles;
use A17\Twill\Repositories\ModuleRepository;
use App\Models\Testingfile;

class TestingfileRepository extends ModuleRepository
{
    use HandleTranslations, HandleSlugs, HandleMedias, HandleFiles;

    public function __construct(Testingfile $model)
    {
        $this->model = $model;
    }
}
`

Last but not least I added this:
in form.blade.php:
`@extends('twill::layouts.form')

@section('contentFields')
    @formField('input', [
        'name' => 'description',
        'label' => 'Description',
        'translated' => true,
        'maxlength' => 100
    ]),
 @formField('files', [
    'name' => 'single_file',
    'label' => 'Single file',
    'note' => 'Add a file'
    ])
@stop
`
Any ideas why the file is not "attached" or properly connected to the module?
fileables table is empty

thanks

**Reply by haringsrob** [reply]
I think you missed one step:

At last, add the filesParams configuration array in your model.

public $filesParams = ['file_role', ...]; // a list of file roles

**Reply by ctocph** [reply]
Hi, yes I'm there now trying to figure out what to add here?

I can't find what it means with "file_role" as it's not obvious to me.

Tried this: public $filesParams = ['file']; but didnt work.

I wish I had a manual covering these options. One could also have wish the twill:make:module would have included it

**Reply by haringsrob** [reply]
It is the same as your field name. I will take note to update the docs :)

**Reply by ctocph** [reply]
yeah now I figured it out - worked good thanks a million @haringsrob
